File: Chapter-05-Linked-Lists/SList-Remove-Back/SList-Remove-Back.py
# Create a function which accepts the head node of a list and removes the last element in the list

# import the list class
import sys
from os.path import dirname, abspath

mainDir = dirname(dirname(abspath(__file__)))
sys.path.append(mainDir)

from singleList import List

# import random module
from random import randrange

# create a new list and populate it
newSList = List()
for i in range(0, 10):
    newSList.add(randrange(100))

# print the existing list
print('The existing list is {} - length is {}').format(newSList.show(), newSList.length())

# function to remove the last element in the list
def removeBack(list, head):
    if head:
        if not head.next:
            list.setHeadNode(None)
            list.setTailNode(None)
        else:
            tempHead = head.next
            trailer  = head
            while tempHead.next:
                trailer = tempHead
                tempHead= tempHead.next
            trailer.next = None
            list.setTailNode(trailer)
        list.recalculateLength()

# print the result of running the remove function
removeBack(newSList, newSList.getHeadNode())
print('The modified list is {} - length is {}').format(newSList.show(), newSList.length())

File: Chapter-05-Linked-Lists/SList-Remove-Back/SList-Remove-Back.js
// Create a function which accepts the head node of a list and removes the last element in the list

// import the list class
var path = require('path'),
    SList= require(path.join(__dirname, '../singleList'))

// create a new list and populate it
var newSList = new SList()
for(let i = 0; i < 10; i++){
  newSList.add(Math.floor(Math.random() * 100))
}

// print the existing list
console.log(`The existing list is ${newSList.show()} - length is ${newSList.length()}`)

// function to remove last element
function removeBack(list, head){
  if (head){
    if (!head.next){
      list.setHeadNode(null)
      list.setTailNode(null)
    } else {
      let tempHead = head.next,
          trailer  = head
      while (tempHead.next){
        trailer = tempHead
        tempHead= tempHead.next
      }
      trailer.next = null
      list.setTailNode(trailer)
    }
    list.recalculateLength()
  }
}

// print the result of running the remove function
removeBack(newSList, newSList.getHeadNode())
console.log(`The modified list is ${newSList.show()} - length is ${newSList.length()}`)
